src/NServiceBus.AcceptanceTests/Outbox/When_headers_contain_special_characters.cs

```diff
@@ -0,0 +1,112 @@
+        public async Task Outbox_should_work()
+        {
+            Requires.OutboxPersistence();
+
+            var context =
+                await Scenario.Define<Context>()
+                    .WithEndpoint<OutboxEndpoint>(b => b.When(session => session.SendLocal(new PlaceOrder())))
+                    .Done(c => c.MessageReceived)
+                    .Run(TimeSpan.FromSeconds(20));
+
+            Assert.IsNotEmpty(context.UnicodeHeaders);
+            CollectionAssert.IsSubsetOf(sentHeaders, context.UnicodeHeaders);
+        }
+
+        class Context : ScenarioContext
+        {
+            public Dictionary<string, string> UnicodeHeaders { get; set; }
+            public bool MessageReceived { get; set; }
+        }
+
+        public class OutboxEndpoint : EndpointConfigurationBuilder
+        {
+            public OutboxEndpoint()
+            {
+                EndpointSetup<DefaultServer>(b =>
+                {
+                    b.EnableOutbox();
+                    b.Pipeline.Register("BlowUpBeforeDispatchBehavior", new BlowUpBeforeDispatchBehavior(), "Force reading the message from Outbox storage.");
+                    b.Recoverability().Immediate(a => a.NumberOfRetries(1));
+                });
+            }
+            class BlowUpBeforeDispatchBehavior : IBehavior<IBatchDispatchContext, IBatchDispatchContext>
+            {
+                public async Task Invoke(IBatchDispatchContext context, Func<IBatchDispatchContext, Task> next)
+                {
+                    if (!SavedOutBoxRecord)
+                    {
+                        SavedOutBoxRecord = true;
+                        throw new Exception();
+                    }
+
+
+                    await next(context).ConfigureAwait(false);
+                }
+            }
+            class PlaceOrderHandler : IHandleMessages<PlaceOrder>
+            {
+                public Task Handle(PlaceOrder message, IMessageHandlerContext context)
+                {
+                    var sendOrderAcknowledgement = new SendOrderAcknowledgement();
+                    var sendOptions = new SendOptions();
+                    sendOptions.RouteToThisEndpoint();
+                    foreach (var header in sentHeaders)
+                    {
+                        sendOptions.SetHeader(header.Key, header.Value);
+                    }
+                    sendOptions.SetMessageId(messageId.ToString());
+                    return context.Send(sendOrderAcknowledgement, sendOptions);
+                }
+            }
+
+            class SendOrderAcknowledgementHandler : IHandleMessages<SendOrderAcknowledgement>
+            {
+                public Context Context { get; set; }
+
+                public Task Handle(SendOrderAcknowledgement message, IMessageHandlerContext context)
+                {
+                    Context.MessageReceived = true;
+                    Context.UnicodeHeaders = (Dictionary<string, string>)context.MessageHeaders;
```

Comment:
doesn't need to be casted afaik. Change the property to ReadOnlyDictionary. The CollectionAssert operation operates on IEnumerables and you can compare a regular Dictionary to a ReadOnlyDictionary.